src/main/java/com/conveyal/gtfs/validator/FlexValidator.java

```diff
@@ -440,62 +574,49 @@ private static boolean hasFareRules(List<FareRule> fareRules, String zoneId) {
     }
 
     /**
-     * Check if a stop id matches any locations or any stop areas.
+     * If a location group id or a location id are not defined, a stop id is required.
      */
-    public static boolean stopIdIsStopAreaOrLocation(
-        String stopId,
-        List<StopArea> stopAreas,
-        List<Location> locations
-    ) {
-        return stopIdIsStopArea(stopId, stopAreas) || stopIdIsLocation(stopId, locations);
+    public static boolean isStopIdRequired(StopTime stopTime) {
+        return stopTime.stop_id == null && (stopTime.location_group_id == null && stopTime.location_id == null);
     }
 
     /**
-     * Check if a stop id matches any stop area, area ids.
+     * If a location group or location is defined, a stop id is forbidden.
      */
-    public static boolean stopIdIsStopArea(String stopId, List<StopArea> stopAreas) {
-        return
-            stopAreas != null &&
-                !stopAreas.isEmpty() &&
-                stopAreas.stream().anyMatch(stopArea -> stopId.equals(stopArea.area_id));
+    public static boolean isStopIdForbidden(StopTime stopTime) {
+        return stopTime.location_group_id != null || stopTime.location_id != null;
```

Comment:
Replace with method `isLocationOrLocationGroupDefined` defined below. You might be able to remove this duplicate method.